src/main/java/tweeting/services/Twitter4JService.java

```diff
@@ -65,10 +80,38 @@ public Twitter4JService(Twitter api) {
         if (StringUtils.isBlank(keyword)) {
             throw new TwitterServiceCallException(MISSING_KEYWORD_MESSAGE);
         }
-        return getHomeTimeline().map(tweets -> tweets.stream()
-                .filter(t -> StringUtils.containsIgnoreCase(t.getMessage(), keyword))
-                .collect(Collectors.toList())
-        );
+        try {
+            // First attempt to retrieve from filtered cache
+            if (filteredCache.containsKeyword(keyword)) {
+                logger.info("Successfully retrieved home timeline filtered by \'" + keyword + "\' from cache.");
+                return Optional.of(filteredCache.getTweets(keyword));
+            }
+            // Next, attempt to retrieve timeline from cache and apply filter
+            else if (homeTimelineCache.getCachedTimeline() != null) {
+                final List<Tweet> tweets = homeTimelineCache.getCachedTimeline()
```

Comment:
inline this section as you are still basically writing this as non streaming code.